### PR TITLE
[Feature] Support setting concurrency for thread pool token

### PR DESCRIPTION
**Pull request by morningman**
## Proposed changes

Now we can submit a group of tasks using thread pool token, and limit
the max concurrency of this task group

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/threadpool.h [hunk 1]
```diff
@@ -362,6 +362,13 @@ class ThreadPoolToken {
     // Returns true if all submissions are complete, false otherwise.
     bool wait_for(const MonoDelta& delta);
 
+    bool need_dispatch();
```
Comment (yangzhg) [inline be/src/util/threadpool.h:365]: need_dispatch is a bit weird. Idle may be better or have other better names

**Comment by weizuo93** [comment]
Based on this patch, base compaction and cumulative compaction could be submitted to thread pool by different tokens. So we could resolve the problem that too many threads are occupied by  base compaction.

**Comment by morningman** [comment]
> Based on this patch, base compaction and cumulative compaction could be submitted to thread pool by different tokens. So we could resolve the problem that too many threads are occupied by base compaction.

Actually, currently implementation of compaction task generation has already reserved at least one thread for cumulative compaction. But base on thread token, it can be implemented easier.

**Review by yangzhg — APPROVED** [review]
LGTM

File: be/src/util/threadpool.h (hunk 1 above)
Comment (morningman) [reply be/src/util/threadpool.h:365]: The purpose of this `need_dispatch` is to express whether the token needs to be put into the queue, so it does not actually mean idle.

**Pull request merged**